### PR TITLE
Feature/yaml selections

### LINKED ISSUE #2172: Advanced node selection syntax
We want to enable a mechanism of node selection that is:

- More powerful, composable, extensible
- In a structured data format
- Possible to check into version control

We think that this is best implemented as YML. It should be similar to CLI `--models` and `--select` syntaxes, but it will also allow us to move beyond what's possible with CLI flags + arguments.

### Selectors

- resource name
- resource type
- model materialization type
- tags
- project/package
- subdirectory
    - file path literals
- Node dependencies. We can make these more verbose in YML than the current selectors on the CLI:
    - parents: `+my_model`
    - children: `my_model+`
    - children and all their parents: `@my_model`
    - *proposed:* models that depend on macros (and their children?): `my_macro+`

### Set logic

- unions (inclusive OR): current default behavior
- exclusion: possible on the CLI with `--exclude`
- intersections (AND): not yet possible, proposed in #2167
- exclusive OR will be possible as the combination of the three above (`union(A,B) —exclude intersect(A,B)`)

### Well defined "pseudo-selectors"

We can encode a dynamic selector that returns resources based on a set of conditions, which dbt uses to pick specific nodes at build time. I'm including a couple possibilities of varying complexity, mainly to spur the imagination:

- `this_package_only`
    - Only execute models + tests that are defined in the current "home? project
    - Dynamic based on the project directory from which it's run.
- `build_if_missing`
    - Exclude model nodes that already exist as relations in the `target` database + schema
- `build_if_changed`
    - point to  `manifest.json` from a different dbt build, and dbt can compare to infer changed resources
    - sensible pattern: select all nodes with changes + their children
- `build_if_updated`
    - point to a `manifest.json` from a different dbt build, _and_ the result of a more recent `dbt source snapshot-freshness`. dbt can determine whether 

### (Very) hypothetical spec

```yml
version: 2

selectors:

  - name: snowplow_marketing_nightly    # human-friendly name for this custom node grouping
    definition:
      - union:          # include nodes for which ANY of the selectors below is true
        - intersect:    # include nodes for which ALL of the selectors below are true
          - tag: nightly
          - tag: marketing
          - package: snowplow
          - materialized: incremental
        - union:
          - resource_name: snowplow_marketing_custom_events
          - file_path: "models/snowplow/marketing/custom_events.sql"
          - model_dir: "snowplow/marketing"
        - intersect:
          - resource_type: seed
          - package: snowplow
          - exclude:
              resource_name: country_codes
  
  - name: ci    # a different custom node grouping
    definition:
      - dynamic: build_if_changed
        parents: false
        children: true
```

```
dbt run --selector snowplow_marketing_nightly
```
```
dbt run --selector ci
dbt test --selector ci
```

### Prior art

This carries on the legacy of several past issues (going back to #550, if not earlier). It's something we've been thinking about for some time.

Looking ahead, I believe that a good approach here will form the basis for features we're very interested in supporting:
- complex workflows
- smarter CI
- dev/prod environments in larger/multi-package projects

### PR DESCRIPTION
**Pull request by beckjake**
resolves #2172 

### Description
This PR implements the advanced node selection syntax, mostly as described in #2172. I've made some changes to accommodate yaml syntax things, but I tried to keep with the spirit of the thing.

Here's a selectors.yml example file (taken from tests):
```yaml
selectors:
  - name: views-or-foos-not-bars
    definition:
      union:
        - method: config.materialized
          value: view
        - tag: foo
        - exclude:
          - tag: bar
``` 

It defines a selector that is the all models materialized as views or tagged `foo`, except without any models tagged `bar`.

I didn't use hologram for parsing here - hologram (well, actually Python's type system) doesn't support recursive type definitions. Hologram also does a poor job of handling ambiguous specifications like this one: it's very hard to get hologram to support both the semi-arbitrary `{'method': 'tag', 'value': 'foo'}` spec and the fully-arbitrary-but-one-key `{'tag': 'foo'}` spec.

#### Selector definitions

There are 3 ways to define a simple selector:
```yaml
definition:
  tag: foo
```
This parses to a list entry containing a dict like `{'tag': 'foo'}`. It's converted, but you can't use modifiers like `@` or `+` here. We could add support for modifiers by examining the keys and values and taking any prefixes/suffixes, or more likely doing `':'.join([key, value])` and passing that in to the string parsing logic. The more I think about this, the more I think it would be good, if only for consistency with the next form.
We could also add support for `exclude` in this syntax if we wanted, though I think it makes the subtle distinction between `tag: foo` and `tag:foo` much more confusing.

```yaml
definition:
  tag:foo
```
This parses to a string entry like `'tag:foo'`, which is then parsed like CLI arguments are. You can use modifiers like `@` or `+` here, though yaml will want you to quote them: (`- "@tag:foo"`)

```yaml
definition:
  method: tag
  value: foo
  childrens_parents: true
```
This parses to a dictionary entry like `{'method': 'tag', 'value': 'foo', 'childrens_parents': True}`. This is what the string form is converted into, and then it goes down the same conversion route.

#### Combinations
Internally, this code still uses the same basic ideas introduced in previous PRs: You can combine values as unions, differences, and intersections of sets of selector definitions. The `union` and `intersection` combinations are themselves selector definitions, and can be used anywhere. Set differences are discussed below, but basically `exclude` can exist anywhere a selector definition could, or within a simple selector definition (which makes it... not so simple).

Only the third form of simple selector definition can be used with exclude. For example, to do what `dbt run --exclude @tag:foo` does:
```yaml
- method: fqn
   value: "*"
   exclude:
   - "@tag:foo"
```

Of course, you can define a one-element union with exclusions if you prefer that syntax:
```yaml
- union:
    - "fqn:*"
    - exclude:
      - "@tag:foo"
```

#### Set differences
The 'exclude' key is the only way to specify set differences. It accepts a list of definitions that are then unioned together. I could definitely be convinced that that's wrong and the value should instead be just a definition. My reasoning rests on the assertion that (at least in Python!) `difference(a, union(b, c, d))` is the same as `difference(a, b, c, d)`, which I feel reasonably confident about.

I think it'd be reasonable to add a `difference` key that acts as an explicit set difference. It would be its own value, as opposed to `exclude`, which I think of as modifying its parents: `{method: 'fqn', value: '*', exclude: ['@tag:foo']}` actually becomes `difference("fqn:*", "@tag:foo")`.

#### Exclude syntax
The syntax isn't immensely satisfying to me, especially around `exclude`. Currently, a union with exclusions has `exclude:` as one of its elements. Does it make more sense for `exclude:` to live on the same level as `union`?:
```yaml
name: my-selector
definition:
  union:
    - tag:foo
    - config.materialized: view
  exclude:
    - tag: bar
```

I don't feel like this is really better at all (I think the lack of indentation makes it hard to parse mentally at a glance), but I don't have great taste on this kind of thing.

### Checklist
 - [X] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
 - [X] I have run this code in development and it appears to resolve the stated issue
 - [X] This PR includes tests, or tests are not required/relevant for this PR
 - [X] I have updated the `CHANGELOG.md` and added information about my change to the "dbt next" section.

### CODE REVIEW

**Comment by mlavoie-sm360** [comment]
Can we make suggestions?

When executing a run using a --select, it would be nice if the logs printed the expanded version of that selector, i.e.

```
selectors:
  - name: my_selector
    definition:
      union:
        - method: tag
          value: something
          parents: true
        - exclude:
            - method: tag
              value: something_else
```

Would then print out something like:

```
dbt run --select my_selector
Running with dbt=0.18.0-rc1
Found 562 models, 578 tests, 8 snapshots, 177 analyses, 323 macros, 4 operations, 40 seed files, 198 sources

15:40:52 | Executing dbt run --models +tag:something --exclude tag:something_else
15:40:52 | 
15:40:52 | Running 2 on-run-start hooks
...
```

**Comment by jtcohen6** [comment]
@mlavoie-sm360 Suggestions very welcome! There's a separate open issue (https://github.com/fishtown-analytics/dbt/issues/2700) around making YAML selectors user-friendlier, specifically for better logging / errors. Would you mind posting your comment there so we don't lose track of it?
